### PR TITLE
Log starttime of dependencies up to the millisecond

### LINKED ISSUE #180: tracked dependencies do not show up correctly in the Application Insights timeline
**Describe the bug**
When I use the Arcus.Observability `LogDependency` extension method on `ILogger`, I've noticed that my tracked dependencies do not show up correctly in the Application Insights timeline:

image
As you can see in the above screenshot, it looks like that my dependency to Timeseries Insights occurred before the initial request to my API was received, which is actually impossible, as my API is calling TSI; so it is not possible that this dependency occurs before the initial request on the API was received.

After some digging around, I have found that it all comes down to the precision of the timestamp that is being logged.  The timestamp of my API request is this:

```
2021-02-24T13:33:26.0349398Z
```

Whereas the timestamp of my tracked TSI dependency is this:

```
2021-02-24T13:33:26Z
```
As you can see, the dependency log's timestamp is only precise to the second, whereas the timestamp of the incoming API request is precise up to the millisecond.
The timestamp of the TSI dependency log will be considered to be 13:33:26.000Z which is then indeed before 13:33:26.0349Z

### PR DESCRIPTION
**Pull request by fgheysels**
When dependencies are tracked, the starttime of the dependency is converted to a string using the InvariantCulture format provider.  However, this provider only converts the datetime to a string with a precision up to the second.
There doesn't seem to be any built-in format provider that converts dates to strings with millisecond precision, so we've specified our own.

closes #180

### CODE REVIEW

**Review by tomkerkhove — APPROVED** [review]

**Comment by fgheysels** [comment]
Don't merge this one yet please (when the tests succeed :) ), I'd like to test / validate this by integrating the preview build in a project.

File: src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/AzureSearchDependencyTests.cs
Comment (tomkerkhove) [inline src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/AzureSearchDependencyTests.cs:33]: Is there a reason why we are changing this?

**Review by tomkerkhove — CHANGES_REQUESTED** [review]
We'll need to update our sink as well:
- Our dependency telemetry converter reads it here: [code](https://github.com/arcus-azure/arcus.observability/blob/master/src/Arcus.Observability.Telemetry.Serilog.Sinks.ApplicationInsights/Converters/DependencyTelemetryConverter.cs#L25)
- That uses  GetAsDateTimeOffset here: [code](https://github.com/arcus-azure/arcus.observability/blob/4115c3d94785b55e388282e1441541f0d4053dcf/src/Arcus.Observability.Telemetry.Serilog.Enrichers/Extensions/LogEventPropertyValueExtensions.cs#L117)

File: src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/AzureSearchDependencyTests.cs
Comment (fgheysels) [reply src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/AzureSearchDependencyTests.cs:33]: For some reason, a lot of integration tests were failing.  It seemed like they were timing out because the dependencies that were written to AppInsights were not retrieved in time.   Sometimes I could reproduce it on my machine, sometimes not.
Then I decided to see what happens if I only retrieve the dependency-logs from AppInsights that occured in the last x minutes, by specifying a timespan filter.  Then, it looks like everything succeeds again.

Offcourse, to do this, I can't rely on random generated timestamps for the events.  I also suspect, but I have no proof for it, that writing a lot of things to appinsights with out-of-order timestamps might hurt AppInsights performance, as I would thing that AppInsights is timeseries based so some re-ordering might happen in appinisghts if you write out of order, but again, I only suspect that.
IMHO, it's also not a big deal to use the current datetime as the timestamp for the event that we're writing; it's actually as random as picking just another time for today's date.

Comment (tomkerkhove) [reply src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/AzureSearchDependencyTests.cs:33]: I think this is related to the fact that we have not updated the application insights sink yet which is not respecting the milliseconds.

Can you test the same after making the changes and also with the original recent offset please?

Comment (fgheysels) [reply src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/AzureSearchDependencyTests.cs:33]: I can't test it using the preview build in my project.  For some reason, it keeps complaining that Arcus.Observability.Correlation when I use the preview build in my project.

Even if I explicitly add all Arcus.Observability projects to my project, it keeps on complaining.
I also have a dependency on Arcus.WEbApi.Logging.Core; maybe they don't play well together ?

```
Could not load file or assembly 'Arcus.Observability.Correlation, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null'. 
The system cannot find the file specified.
```

Comment (fgheysels) [reply src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/AzureSearchDependencyTests.cs:33]: If I remove the Arcus.WebApi.Logging package from my project, I don't have the AssemblyLoadException anymore.

Comment (fgheysels) [reply src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/AzureSearchDependencyTests.cs:33]: > I think this is related to the fact that we have not updated the application insights sink yet which is not respecting the milliseconds.
> 
> Can you test the same after making the changes and also with the original recent offset please?

I've tested the current preview build in my project, and the timestamp of the dependency is correct up to the millisecond.
However, I see that the timestamps of all other traces in AppInsights have an even higher precision (it's in ticks).  Think we should also go to that precision.  I'll do some updates here.

Comment (fgheysels) [reply src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/AzureSearchDependencyTests.cs:33]: > I think this is related to the fact that we have not updated the application insights sink yet which is not respecting the milliseconds.

There are tests available which test if the date that is transformed by the appinsights converter is set correct.

File: src/Arcus.Observability.Telemetry.AspNetCore/Extensions/ILoggerExtensions.cs
Comment (tomkerkhove) [inline src/Arcus.Observability.Telemetry.AspNetCore/Extensions/ILoggerExtensions.cs:15]: Since we use it multiple times we should maybe move it to a shared place?

File: src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/ApplicationInsightsSinkTests.cs
Comment (tomkerkhove) [inline src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/ApplicationInsightsSinkTests.cs:66]: Can we revert these changes?

**Review by tomkerkhove — APPROVED** [review]

File: src/Arcus.Observability.Telemetry.AspNetCore/Extensions/ILoggerExtensions.cs
Comment (fgheysels) [reply src/Arcus.Observability.Telemetry.AspNetCore/Extensions/ILoggerExtensions.cs:15]: Yes, I was thinking of that as well.   Problem is where to put it ?  I initially thought to put it in Telemetry.Core, but Telemetry.AspNetCore has no dependency on Telemetry.Core, and I thought it would maybe be overkill to create that dependency, but we can do that

Comment (tomkerkhove) [reply src/Arcus.Observability.Telemetry.AspNetCore/Extensions/ILoggerExtensions.cs:15]: Let's do that!

File: src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/ApplicationInsightsSinkTests.cs
Comment (fgheysels) [reply src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/ApplicationInsightsSinkTests.cs:66]: Yes we can; I just thought it was overkill to query AppInsights every second.

Comment (tomkerkhove) [reply src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/ApplicationInsightsSinkTests.cs:66]: This seems like a long wait though?
https://github.com/arcus-azure/arcus.observability/pull/181/files/6c7cfe05d76513b14edd981cb8fe4602432ffbe1#diff-27f08c3b8059324291530235f56cd9bb92ef433ef8842cd03175f58a2213f06bR68

File: src/Arcus.Observability.Telemetry.AspNetCore/Extensions/ILoggerExtensions.cs
Comment (fgheysels) [reply src/Arcus.Observability.Telemetry.AspNetCore/Extensions/ILoggerExtensions.cs:15]: My bad, the dependency was already there.

**Comment by fgheysels** [comment]
> We'll need to update our sink as well:
> 
> * Our dependency telemetry converter reads it here: [code](https://github.com/arcus-azure/arcus.observability/blob/master/src/Arcus.Observability.Telemetry.Serilog.Sinks.ApplicationInsights/Converters/DependencyTelemetryConverter.cs#L25)
> * That uses  GetAsDateTimeOffset here: [code](https://github.com/arcus-azure/arcus.observability/blob/4115c3d94785b55e388282e1441541f0d4053dcf/src/Arcus.Observability.Telemetry.Serilog.Enrichers/Extensions/LogEventPropertyValueExtensions.cs#L117)

I don't think we need to change anything to the DependencyTelemetryConverter.  When you feed a string representation of a datetimeoffset with ticks, they are converted correctly.  Lets see if I can add an additional test for it.

File: src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/ApplicationInsightsSinkTests.cs
Comment (fgheysels) [reply src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/ApplicationInsightsSinkTests.cs:66]: That's the max timeout.  So if it succeeds before that time, it won't wait that long

File: src/Arcus.Observability.Telemetry.AspNetCore/Extensions/ILoggerExtensions.cs
Comment (fgheysels) [reply src/Arcus.Observability.Telemetry.AspNetCore/Extensions/ILoggerExtensions.cs]: Done.

**Comment by fgheysels** [comment]
> > We'll need to update our sink as well:
> > 
> > * Our dependency telemetry converter reads it here: [code](https://github.com/arcus-azure/arcus.observability/blob/master/src/Arcus.Observability.Telemetry.Serilog.Sinks.ApplicationInsights/Converters/DependencyTelemetryConverter.cs#L25)
> > * That uses  GetAsDateTimeOffset here: [code](https://github.com/arcus-azure/arcus.observability/blob/4115c3d94785b55e388282e1441541f0d4053dcf/src/Arcus.Observability.Telemetry.Serilog.Enrichers/Extensions/LogEventPropertyValueExtensions.cs#L117)
> 
> I don't think we need to change anything to the DependencyTelemetryConverter. When you feed a string representation of a datetimeoffset with ticks, they are converted correctly. Lets see if I can add an additional test for it.

Added an additional test, but I'm not that satisfied by that test actually.  Don't know if we can write it in another way

File: src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/ApplicationInsightsSinkTests.cs
Comment (fgheysels) [reply src/Arcus.Observability.Tests.Integration/Serilog/Sinks/ApplicationInsights/ApplicationInsightsSinkTests.cs]: Put it back to the original (7 minutes).  Kept the 3 seconds delay though instead of the 1 sec delay between app insights queries

**Comment by fgheysels** [comment]
> Don't merge this one yet please (when the tests succeed :) ), I'd like to test / validate this by integrating the preview build in a project.

Tested it, and it is ok (although I had to remove the WebApi.Logging package from my project to be able to test the preview nuget package from myget).

**Pull request merged**
